packages/decoder/src/union.ts

```diff
@@ -0,0 +1,26 @@
+import * as PR from "@effect/schema/ParseResult"
+import type { ReadonlyArray } from "effect"
+import { Effect, Either } from "effect"
+
+import type { Decoder, OutputOf } from "@typed/decoder/decoder"
+
+export function union<Members extends ReadonlyArray.NonEmptyReadonlyArray<Decoder<unknown, any>>>(
+  ...members: Members
+): Decoder<unknown, OutputOf<Members[number]>> {
+  return (i, options) =>
+    Effect.gen(function*($) {
+      const failures: Array<PR.ParseErrors> = []
+
+      for (const member of members) {
+        const result = yield* $(Effect.either(member(i, options)))
+
+        if (Either.isRight(result)) {
```

Comment:
I've added the push.

Do you think there would be a way to leverage the Schema union function? - it would be nice if the error results were the same in both libraries.

Reply:
I agree it'd be ideal for them to have the same error messages. We won't be able to convert a Decoder back into a Schema to utilize their public API unless they are `SchemaDecoder` instances. We could probably add support for more information on individual decoders, e.g. `StructDecoder<Props> { readonly props: Props, ... }` and attempt to utilize that information to build out something closer to the code from Schema here - https://github.com/Effect-TS/schema/blob/main/src/Parser.ts#L818